functions/db/deliveryRequests/onCreate.f.js

```diff
@@ -0,0 +1,134 @@
+const functions = require('firebase-functions');
+const moment = require('moment');
+const enums = require('../../Enums.js');
+
+/*
+ * When a delivery request is created, send notifications to 
+ * receiving agencies:
+ *     If one specific RA is requested, send notification regardless 
+ *     of availability.
+ *     Otherwise, send notifications to all available RAs in the
+ *     pending list.
+ */
+exports = module.exports = functions.database
+    .ref('/delivery_requests/{pushId}')
+    .onCreate(event => {
+        // TODO: setup Admin SDK in the future?
+        const requestRef = event.data.ref;
+        const accountsRef = requestRef.parent.parent.child('accounts');
+        const dasRef = accountsRef.parent.child('donating_agencies');
+
+        var requestKey = event.params.pushId;
+        var request = event.data.val();
+        var raInfo = request.receivingAgency;
+        console.info('New recurring request added: ' + requestKey);
+        
+        if (request.status !== enums.RequestStatus.PENDING) {
+            console.error('ERROR: Request should have pending status upon request.');
+            return Promise.resolve();
+        }
+
+        if (raInfo.claimed) {
+            console.error('ERROR: RA shouldn\'t be confirmed upon request.');
+            return Promise.resolve();
+        }
+
+        if (!raInfo.requested && !raInfo.pending) {
+            console.error('ERROR: No RAs in the request to notify.');
+            return Promise.resolve();
+        }
+
+        // create pickup request notification
+        var notification = {
+            type: enums.NotificationType.RECURRING_PICKUP_REQUEST,
+            content: requestKey
+        };
+
+        // If a specific RA was requested, force push notification
+        if (raInfo.requested) {
+            console.info('A specific RA requested: ' + raInfo.requested);
+            var raRef = accountsRef.child(raInfo.requested);
+
+            return pushNotification(raRef, notification, 'RA');
+        }
+
+        console.info('No specific RA requested, ' + raInfo.pending.length + 
+            ' pending RAs');
+
+        // Compose all promises to get RA snapshots from db
+        var raSnapPromises = [];
+        for (let ra in raInfo.pending) {
+            raSnapPromises.push(getRASnapPromise(accountsRef, raInfo.pending[ra]));
+        }
+
+        // Get all RA snapshot request results and process them
+        return Promise.all(raSnapPromises).then((results) => {
+            var promise = null;
+
+            for (let i in results) {
+                var raSnap = results[i];
+                var available = isAvailable(raSnap.val(), request);
+
+                console.info('RA "' + raSnap.key + '": available=' + available);
+
+                if (available) {
+                    promise = pushNotification(raSnap.ref, notification, 'RA');
```

Comment:
This promise will be just for the last RA, right? It should be for all that have been notified I think...

Reply:
Yeah it's just the last RA. There are 2 purposes here why we're keeping the `promise`:

1. In Cloud Functions, we need to return a `Promise` otherwise the server might terminate the instance immaturely before the async function has finished executing and we could lose data.
2. For the check below to see if we ever processed any RA.

However, you might be right. I'm still not exactly sure how it works, but it's probably possible if I only return the `promise` for the last RA, the function would only wait until the write for that RA is done? I'll change it to keeping a list and return `Promise.all()`!